engines/wintermute/base/gfx/opengl/base_render_opengl3d_shader.cpp

```diff
@@ -0,0 +1,607 @@
+	setProjection2D();
+
+	Math::Vector4d color;
+	color.x() = r / 255.0f;
+	color.y() = g / 255.0f;
+	color.z() = b / 255.0f;
+	color.w() = a / 255.0f;
+
+	glDisable(GL_DEPTH_TEST);
+	glEnable(GL_BLEND);
+	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+	glBindTexture(GL_TEXTURE_2D, 0);
+	glBindBuffer(GL_ARRAY_BUFFER, _fadeVBO);
+
+	_fadeShader->use();
+	_fadeShader->setUniform("color", color);
+	_fadeShader->setUniform("projMatrix", _projectionMatrix2d);
+
+	glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+
+	setup2D(true);
+}
+
+bool BaseRenderOpenGL3DShader::drawLine(int x1, int y1, int x2, int y2, uint32 color) {
+	warning("BaseRenderOpenGL3DShader::drawLine not yet implemented");
+	return true;
+}
+
+bool BaseRenderOpenGL3DShader::drawRect(int x1, int y1, int x2, int y2, uint32 color, int width) {
+	warning("BaseRenderOpenGL3DShader::drawRect not yet implemented");
+	return true;
+}
+
+bool BaseRenderOpenGL3DShader::setProjection() {
+	// is the viewport already set here?
+	float viewportWidth = _viewportRect.right - _viewportRect.left;
+	float viewportHeight = _viewportRect.bottom - _viewportRect.top;
+
+	float verticalViewAngle = _fov;
+	float aspectRatio = float(viewportWidth) / float(viewportHeight);
+	// same defaults as wme
+	float nearPlane = 90.0f;
+	float farPlane = 10000.0f;
+
+	float top = nearPlane * tanf(verticalViewAngle * 0.5f);
+
+	_projectionMatrix3d = Math::makeFrustumMatrix(-top * aspectRatio, top * aspectRatio, -top, top, nearPlane, farPlane);
+	return true;
+}
+
+bool BaseRenderOpenGL3DShader::setProjection2D() {
+	float viewportWidth = _viewportRect.right - _viewportRect.left;
+	float viewportHeight = _viewportRect.bottom - _viewportRect.top;
+
+	float nearPlane = -1.0f;
+	float farPlane = 100.0f;
+
+	_projectionMatrix2d.setToIdentity();
+
+	_projectionMatrix2d(0, 0) = 2.0f / viewportWidth;
+	_projectionMatrix2d(1, 1) = 2.0f / viewportHeight;
+	_projectionMatrix2d(2, 2) = 2.0f / (farPlane - nearPlane);
+
+	_projectionMatrix2d(3, 0) = -1.0f;
+	_projectionMatrix2d(3, 1) = -1.0f;
+	_projectionMatrix2d(3, 2) = -(farPlane + nearPlane) / (farPlane - nearPlane);
+
+	_shadowMaskShader->use();
+	_shadowMaskShader->setUniform("projMatrix", _projectionMatrix2d);
+	return true;
+}
+
+void BaseRenderOpenGL3DShader::resetModelViewTransform() {
+	glMatrixMode(GL_MODELVIEW);
+	glLoadIdentity();
+}
+
+void BaseRenderOpenGL3DShader::setWorldTransform(const Math::Matrix4 &transform) {
+	Math::Matrix4 tmp = transform;
+	tmp.transpose();
+
+	Math::Matrix4 newInvertedTranspose = tmp * _lastViewMatrix;
+	newInvertedTranspose.inverse();
+	newInvertedTranspose.transpose();
+
+	_modelXShader->use();
+	_modelXShader->setUniform("modelMatrix", tmp);
+	_modelXShader->setUniform("normalMatrix", newInvertedTranspose);
+
+	_shadowVolumeShader->use();
+	_shadowVolumeShader->setUniform("modelMatrix", tmp);
+}
+
+bool BaseRenderOpenGL3DShader::windowedBlt() {
+	warning("BaseRenderOpenGL3DShader::windowedBlt not yet implemented");
```

Comment:
there are severals not implemented functions, what is a plan to implement them?

Reply:
This particular candidate here just calls flip() in the original WME. For the other ones it depends I guess. Some functions are related to displaying indicator screens of savegames it seems, I have not looked into how this works and would prefer to consider this out of scope for now. Some functions are related to non-stencil shadows, if there is time, they could be implemented, it should not be too much work. Also there is some missing 2d stuff in drawSpriteEx(..), among the non implemented stuff, I would like to give this priority since I already know one (2d) game which uses this.